media-proxy/include/mesh/concurrency.h

```diff
@@ -0,0 +1,215 @@
+
+    Context();
+
+    friend Context& Background();
+    friend class WithCancel;
+};
+
+/**
+ * mesh::context::Background()
+ * 
+ * Returns an empty Context with cancellation. It is typically used in the
+ * main function and initialization. It should be the parent context for
+ * context::WithCancel and context::WithTimeout.
+ */
+Context& Background();
+
+/**
+ * mesh::context::WithCancel
+ * 
+ * Creates a new Context with cancellation. When the parent context is
+ * cancelled, it triggers cancellation of the child context and all descending
+ * child contexts in the chain. On the other hand, if the context is cancelled
+ * by calling stop(), the parent context is not affected because the
+ * cancellation signal propagates only down to the child contexts in the chain.
+ * 
+ * This concept enables propagation of the cancellation signal from the top
+ * parent context down to the very bottom child context and stopping all
+ * threads and blocking I/O calls that depend on any context in the chain.
+ */
+class WithCancel : public Context {
+public:
+    WithCancel(Context& parent);
+
+private:
+    Context& parent;
+};
+
+/**
+ * mesh::context::WithTimeout
+ * 
+ * Creates a new Context with cancellation and a timeout. The time interval
+ * starts to count down at creation. When the time is out, the context is
+ * cancelled, which triggers cancellation of all child contexts in the chain.
+ */
+class WithTimeout : public WithCancel {
+public:
+    WithTimeout(Context& parent, std::chrono::milliseconds timeout_ms);
+    ~WithTimeout() override;
+
+private:
+    std::future<void> async_cb;
+};
+
+} // namespace context
+
+namespace thread {
+
+/**
+ * Defer
+ * 
+ * Allows to register a callback to be called when the defer object leaves
+ * the scope of visibility. Deferred callbacks can be used for automatic
+ * deallocation of resources created in a function when the function exits.
+ * Example:
+ * 
+ * void func() {
+ *     printf("Enter the function\n");
+ *     ...
+ *     Defer d1([]{ printf("Deferred action 1"); });
+ *     ...
+ *     Defer d2([]{ printf("Deferred action 2"); });
+ *     ...
+ *     printf("Exit the function\n");
+ * }
+ * 
+ * Output:
+ *     Enter the function
+ *     Exit the function
+ *     Deferred action 2
+ *     Deferred action 1
+ */
+class Defer {
+public:
+    Defer(std::function<void()> callback) : cb(callback) {}
+    ~Defer() { if (cb) cb(); }
+
+private:
+    std::function<void()> cb;
+};
+
+/**
+ * Sleep
+ * 
+ * Sleeps for the given interval of time, which interrupts immediately when
+ * the context is cancelled.
+ */
+void Sleep(context::Context& ctx, std::chrono::milliseconds interval_ms);
+
+/**
+ * Channel
+ * 
+ * A thread-safe queue template supporting cancellation of blocking calls
+ * put() and get() by checking context::Context. To set a timeout on the
+ * blocking calls, context::WithTimeout is supposed to be used.
+ * Example:
+ * ...
+ */
+template<typename T>
+class Channel {
```

Comment:
what is a use case for the channel in our project? Do you want to pass pointer to buffers between connections?

Reply:
Channels are thread safe queues. They will be used to deliver events within Media Proxy:
* Connection created
* Connection destroyed
* Bridge created
* Bridge destroyed
* Multipoint Group created
* Multipoint Group destroyed
* Other events related to transitions between internal states.
Another uses case for Channel is a wait group (see `sync.WaitGroup` in Golang). It allows to control the flow and wait until some processing in parallel threads is completed. I'll show it later in code samples.

Reply:
What if we implement it as a lock free queue? (for example with [this](https://www.boost.org/doc/libs/1_86_0/doc/html/lockfree.html))  To be honest, I'm not sure how this will affect performance and whether we need good performance in events at all, but I just want to throw my thought out there.

Reply:
Events will never be generated in the hot path of Data Plane. We don't need tangible performance in Control Plane at all. The idea of a lock-free queue is good, however it will be way more complex because we can't combine the `condition_variable` waiting with lock-free primitives like atomics. We can't use the C++20 `wait` of atomics because we can't cancel it. I even can't measure the complication that the lock-free implementation could bring here.

So, let's don't optimize prematurely ;-)

Reply:
agree, lack of `wait` would be an issue.